Sources/SemanticIndex/SemanticIndexManager.swift

```diff
@@ -683,34 +687,64 @@ package final actor SemanticIndexManager {
     // We will check the up-to-date status again in `IndexTaskDescription.execute`. This ensures that if we schedule
     // schedule two indexing jobs for the same file in quick succession, only the first one actually updates the index
     // store and the second one will be a no-op once it runs.
-    let outOfDateFiles = await filesToIndex(toCover: files, indexFilesWithUpToDateUnits: indexFilesWithUpToDateUnit)
+    var filesToIndex = await filesToIndex(toCover: files, indexFilesWithUpToDateUnits: indexFilesWithUpToDateUnit)
       // sort files to get deterministic indexing order
-      .sorted(by: { $0.sourceFile.stringValue < $1.sourceFile.stringValue })
-
-    if outOfDateFiles.isEmpty {
-      // Early exit if there are no files to index.
-      return Task {}
-    }
-
-    logger.debug("Scheduling indexing of \(outOfDateFiles.map(\.sourceFile.stringValue).joined(separator: ", "))")
-
-    // Sort the targets in topological order so that low-level targets get built before high-level targets, allowing us
-    // to index the low-level targets ASAP.
-    var filesByTarget: [BuildTargetIdentifier: [FileToIndex]] = [:]
+      .sorted(by: { $0.file.sourceFile.stringValue < $1.file.sourceFile.stringValue })
 
     // The number of index tasks that don't currently have an in-progress task associated with it.
     // The denominator in the index progress should get incremented by this amount.
     // We don't want to increment the denominator for tasks that already have an index in progress.
     var newIndexTasks = 0
-    for file in outOfDateFiles {
-      if inProgressIndexTasks[file] == nil {
+    var alreadyScheduledTasks: Set<FileToIndex> = []
+    for file in filesToIndex {
+      let inProgress = inProgressIndexTasks[file.file]
+
+      let shouldScheduleIndexing: Bool
+      switch inProgress?.state {
+      case nil:
         newIndexTasks += 1
+        shouldScheduleIndexing = true
+      case .creatingIndexTask, .waitingForPreparation:
+        // We already have a task that indexes the file but hasn't started preparation yet. Indexing the file again
+        // won't produce any new results.
+        alreadyScheduledTasks.insert(file.file)
+        shouldScheduleIndexing = false
+      case .preparing(_, _), .updatingIndexStore(_, _):
```

Comment:
If it's currently preparing, we don't need to index it again do we? An update to the file itself shouldn't cause its own preparation to change. One possible exception to that could be a header file though.

Reply:
We could have triggered re-indexing because build settings (or built graph in general) have changed, in which case we might also need to re-prepare, eg. to pick up new dependencies.